### PR TITLE
Use onBeforeCompile to customize shader for instancing

### PR DESCRIPTION
**Pull request by Oletus**
Replacing shaders in THREE.ShaderLib is a bad practice, as it interferes with normal operation of built-in materials. When materials need to be customized, it's better to use onBeforeCompile.

### CODE REVIEW

**Review by Mugen87 — APPROVED** [review]

File: examples/webgl_buffergeometry_instancing_lambert.html [hunk 1]
```diff
@@ -292,15 +285,16 @@
...
 
 
 			// custom depth material - required for instanced shadows
 
-			var shader = THREE.ShaderLib[ 'customDepthRGBA' ];
-
-			var uniforms = THREE.UniformsUtils.clone( shader.uniforms );
+			var uniforms = THREE.UniformsUtils.clone( customDepthToRGBAShader.uniforms );
 
 			var customDepthMaterial = new THREE.ShaderMaterial( {
```
Comment (sciecode) [inline examples/webgl_buffergeometry_instancing_lambert.html]: Custom shader can also be replaced with `THREE.MeshDepthMaterial` and explicitly modified with `onBeforeCompile` like the other suggestion.

**Review by sciecode — COMMENTED** [review]
If we're no longer modifying `ShaderLib` we might as well hide the unnecessary shader code.

File: examples/webgl_buffergeometry_instancing_lambert.html [hunk 2]
```diff
@@ -292,15 +285,16 @@
 
 			} );
 
+			material.onBeforeCompile = function( shader ) {
+				shader.vertexShader = customLambertVertexShader;
```
[the remark this answers is not in the document]
Comment (WestLangley) [reply examples/webgl_buffergeometry_instancing_lambert.html]: I created the example this way so users could clearly see the differences without having to assemble the shader chunks manually to understand the overall logic.

[the remark this answers is not in the document]
Comment (sciecode) [reply examples/webgl_buffergeometry_instancing_lambert.html]: Well, using `vertexShader.replace(<chunk>, ...)` you can still clearly see the differences, without needing to copy the entire shader. But I can see the reason for wanting to keep it this way.

[the remark this answers is not in the document]
Comment (WestLangley) [reply examples/webgl_buffergeometry_instancing_lambert.html]: I'll rephrase it to what I meant to say...

I created the example this way so users could easily read the shader without having to assemble the shader chunks manually.

File: examples/webgl_buffergeometry_instancing_lambert.html (hunk 1 above)
Comment (Oletus) [reply examples/webgl_buffergeometry_instancing_lambert.html]: Done!

Comment (WestLangley) [reply examples/webgl_buffergeometry_instancing_lambert.html]: So, now, what is the point of `#ifdef INSTANCED`?

Comment (Oletus) [reply examples/webgl_buffergeometry_instancing_lambert.html]: True, it's not really needed anymore, removed it.

Comment (WestLangley) [reply examples/webgl_buffergeometry_instancing_lambert.html]: Do you think it is still easy to see the changes required to support instancing?

Comment (sciecode) [reply examples/webgl_buffergeometry_instancing_lambert.html]: I think so. I would just add the missing `//instanced` on `LambertVertexShader` and it should still be clear what are the changes necessary.

Comment (Oletus) [reply examples/webgl_buffergeometry_instancing_lambert.html]: Added the // instanced comment.

**Comment by Mugen87** [comment]
@mrdoob This one looks good!

**Comment by WestLangley** [comment]
Thanks!

**Comment by mrdoob** [comment]
Thanks!
